vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/pipeline_util.cc

```diff
@@ -0,0 +1,220 @@
+  }
+
+  std::string parsed_classifier_type = *classifier_type;
+
+  if (parsed_classifier_type.compare("LINEAR")) {
+    return base::nullopt;
+  }
+
+  base::Value* specified_classes = classifier_value->FindListKey("classes");
+  if (!specified_classes) {
+    return base::nullopt;
+  }
+
+  auto specified_classes_list = specified_classes->GetList();
+  for (size_t i = 0; i < specified_classes_list.size(); i++) {
+    const base::Value& class_name = specified_classes_list[i];
+    classes.push_back(class_name.GetString());
+  }
+
+  base::Value* class_weights = classifier_value->FindDictKey("class_weights");
+  if (!class_weights) {
+    return base::nullopt;
+  }
+
+  std::map<std::string, VectorData> weights;
+  for (size_t i = 0; i < classes.size(); i++) {
+    base::Value* this_class = class_weights->FindListKey(classes[i]);
+    if (!this_class) {
+      return base::nullopt;
+    }
+    std::vector<double> class_coef_weights;
+    auto this_class_list = this_class->GetList();
+    for (size_t j = 0; j < this_class_list.size(); j++) {
+      const base::Value& weight = this_class_list[j];
+      class_coef_weights.push_back(weight.GetDouble());
+    }
+    weights[classes[i]] = VectorData(class_coef_weights);
+  }
+
+  std::map<std::string, double> specified_biases;
+  base::Value* biases = classifier_value->FindListKey("biases");
+  if (!biases) {
+    return base::nullopt;
+  }
+
+  auto biases_list = biases->GetList();
+  if (biases_list.size() != classes.size()) {
+    return base::nullopt;
+  }
+
+  for (size_t i = 0; i < biases_list.size(); i++) {
+    const base::Value& this_bias = biases_list[i];
+    specified_biases[classes[i]] = this_bias.GetDouble();
+  }
+
+  base::Optional<model::Linear> linear_model =
+      model::Linear(weights, specified_biases);
+  return linear_model;
+}
+
+base::Optional<PipelineInfo> ParsePipelineJSON(const std::string& json) {
+  base::Optional<base::Value> root = base::JSONReader::Read(json);
+
+  if (!root) {
+    return base::nullopt;
+  }
+
+  base::Optional<int> version_value = root->FindIntKey("version");
+  if (!version_value.has_value()) {
+    return base::nullopt;
+  }
+  int version_number = version_value.value();
+  uint16_t version = static_cast<uint16_t>(version_number);
+
+  std::string* timestamp_value = root->FindStringKey("timestamp");
+  if (!timestamp_value) {
+    return base::nullopt;
+  }
+  std::string timestamp = *timestamp_value;
+
+  std::string* locale_value = root->FindStringKey("locale");
+  if (!locale_value) {
+    return base::nullopt;
+  }
+  std::string locale = *locale_value;
+
+  base::Value* transformations_value = root->FindKey("transformations");
+  base::Optional<TransformationVector> transformations_opt =
+      ParseTransformationsJSON(transformations_value);
+  if (!transformations_opt.has_value()) {
+    return base::nullopt;
+  }
+
+  base::Value* classifier_value = root->FindKey("classifier");
+  const base::Optional<model::Linear> linear_model_opt =
```

Comment:
`opt` abbreviation, do we need the notation?

Reply:
Done. Replaced with `optional`.